source/common/http/sidestream_watermark.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "envoy/http/async_client.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * Sidestream watermark callback implementation for stream filter that either handles decoding only
+ * or handles both encoding and decoding.
+ */
+class StreamFilterSidestreamWatermarkCallbacks : public Http::SidestreamWatermarkCallbacks {
+public:
+  void onSidestreamAboveHighWatermark() final {
+    // Sidestream push back downstream, if callback is configured.
+    if (decode_callback_ != nullptr) {
+      decode_callback_->onDecoderFilterAboveWriteBufferHighWatermark();
+    }
+
+    // Sidestream push back upstream, if callback is configured.
+    if (encode_callback_ != nullptr) {
+      encode_callback_->onEncoderFilterAboveWriteBufferHighWatermark();
+    }
+  }
+
+  void onSidestreamBelowLowWatermark() final {
+    if (decode_callback_ != nullptr) {
+      decode_callback_->onDecoderFilterBelowWriteBufferLowWatermark();
+    }
+
+    if (encode_callback_ != nullptr) {
+      encode_callback_->onEncoderFilterBelowWriteBufferLowWatermark();
+    }
+  }
+
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks* decode_callback) {
```

Comment:
if you're allowing these to be set/unset at arbitrary points I think you have to track state when they're not set. 
i.e. handle  onSidestreamBelowLowWatermark() being called, then setDecoderFilterCallbacks() called
I think it'd be easier to set in the constructor if possible?

Reply:
My initial design of this class is a simple wrapper class that doesn't manage/own the resource, so I followed The Rule of 0 that just uses compiler-generated special member functions lie c'tor

Specific ordering between `onWatermark ` and `setFilterCallbacks` is not enforced by code but the correct usage of this class, as you pointed out [below](https://github.com/envoyproxy/envoy/pull/35077#discussion_r1686916453), is: `setFilterCallbacks` before stream is established 

IIUC, adding some code comments as per [this review](https://github.com/envoyproxy/envoy/pull/35077#discussion_r1686916453) will also address the comments here. If not, please let me know